### ISSUE TITLE
Shouldn't Time32 and Time64 objects return datetime64 type?

### ISSUE THREAD

**Issue by jondoesntgit** [PyTables/PyTables#820]
Consider the sample script:

```python
import tables as pt
import time

h5file = pt.open_file('tmp.h5', 'w')

class Description(pt.IsDescription):
    timestamp32 = pt.Time32Col()
    timestamp64 = pt.Time64Col()
    
table = h5file.create_table(h5file.root, 'sample', Description)

N = 10
for i in range(N):
    row = table.row
    row['timestamp32'] = time.time()
    row['timestamp64'] = time.time()
    row.append()
table.flush()
```

Now, if I want to get my data out, I can run

```python
timestamp32 = table.cols.timestamp32[:]
timestamp64 = table.cols.timestamp64[:]
```

But then I'm left with numpy arrays of type `int32` and `float64` respectively, even though the underlying types in hdf5 are `time32` and `time64` (c.f., `table.cols.timestamp32[:].type`). 

The behavior that I was expecting can be achieved with the following:

```python
timestamp32 = table.cols.timestamp32[:].astype('datetime64[s]')
timestamp64 = (table.cols.timestamp64[:] * 1e9).astype('datetime64[ns]')
```

Is there any reason why this can't be the default behavior in pytables? Perhaps compatibility with pandas?

**Comment by jondoesntgit** [comment]
Also, I notice that in my `Description` class, I get the expected behavior for default values for a Float32Col with nan values, but not with a Time64Col.

```python
class Description(pt.IsDescription):
    timestamp32 = pt.Time32Col(dflt=-1) # works as expected
    timestamp64 = pt.Time64Col(dflt=np.nan) # does not work, fills with zeros 
    measurement = pt.Float32Col(dflt=np.nan) # works as expected
```

**Comment by jondoesntgit** [comment]
Maybe some sort of default implementation with the `_convert_time64 function`?

**Comment by avalentino** [comment]
Hi @jondoesntgit 

> Is there any reason why this can't be the default behavior in pytables?

numpy datetime did not exist (or was just experimental) when PyTables TimeXX types were created.
After that I think that no one had the time and motivation to improve things.
IMHO a better integration with numpy datetime types would be very nice to have.
Anyway the management of the [datetime units](https://numpy.org/doc/stable/reference/arrays.datetime.html#datetime-units) shall be done properly and the choice of a default is not obvious.

I have found a somewhat correlated issue: #284.

Regarding ``dflt=np.nan``, yes it is an issue.
Please refer to #410 for any further discussion.

**Comment by jondoesntgit** [comment]
I'm back with a progress report after working on this for two days:

It appears that the Time32 and Time64 specs in the HDF5 library are no longer maintained. H5Py doesn't support these data types (raises a `TypeError: No NumPy equivalent for TypeTimeID exists` if you try to open a PyTable HDF5 file with those datatypes). `h5dump` calls those data types `not yet implemented`. I think `h5ls` has trouble with it as well.

If we were to further work on this, then PyTables would be the only major library that supports these data types. And with 32-bit precision for the integer number of seconds since the 1970 epoch, we run into an obsolescence problem in 2038.

**Comment by jondoesntgit** [comment]
As I'm thinking about this, I think the best way to get such a feature moving forward is to create new data types, rather than trying to duck type the existing Time32 and Time64 data types (this would almost certainly break existing code). I think the first step would be opening up atom.py and adding in something like DateTime64Atom(DateTimeAtom). A default resolution of 'ns' would give a range between [ 1678 AD, 2262 AD], though for some applications, they may require much finer precision.

**Comment by avalentino** [comment]
Thanks @jondoesntgit.
I agree with you on the approach of having a new type instead of changing the existing ones.
Regarding the default resolution my suggestion is to not have a default at all and force the user to specify it at creation time. This is the numpy approach if I'm not wrong.

**Comment by jondoesntgit** [comment]
Perhaps by specifying a unit attribute, as suggested in #821.

**Comment by avalentino** [comment]
> Perhaps by specifying a unit attribute, as suggested in #821.

yes, the general mechanism for supporting features that are not directly supported by HDF5 is to use "system attributes" to store the missing piece of information.
